src/pudl/workspace/setup.py

```diff
@@ -1,15 +1,97 @@
 """Tools for setting up and managing PUDL workspaces."""
 import importlib
+import os
 import pathlib
 import shutil
+from pathlib import Path
+from typing import IO
 
 import yaml
+from dotenv import load_dotenv
 
 import pudl.logging_helpers
 
 logger = pudl.logging_helpers.get_logger(__name__)
 
 
+def get_defaults(
+    input_dir: str | None = None,
+    output_dir: str | None = None,
+    yaml_file: IO | None = None,
+    default_pudl_yaml: Path | None = Path.home() / ".pudl.yml",
+) -> dict[str, str]:
+    """Derive PUDL workspace paths from specified input/output directories.
+
+    Determines input/output directory locations from YAML, then overrides with
+    env vars, then overrides with keywords passed in.
+
+    Input/output workspace roots can be the same directories.
+
+    Note: will update PUDL_OUTPUT and PUDL_INPUT env vars if they are
+    overridden by kwargs, so that Dagster configurations in child processes see
+    the updated configs when they read from env vars.
+
+    Args:
+        input_dir: equivalent to PUDL_INPUT environment variable, but overrides
+            that value. Derived paths treat the parent directory as the input
+            workspace root.
+        output_dir: equivalent to PUDL_OUTPUT environment variable, but
+            overrides that value. Derived paths treat the parent directory as
+            the output workspace root.
+        yaml_file: a buffer including the YAML configuration. The `pudl_in` and
+            `pudl_out` keys within this file correspond to the input/output
+            workspace roots directly, instead of through parents.
+
+    Returns:
+        dictionary with a variety of different paths where inputs/outputs are
+        to be found.
+    """
+    load_dotenv()
+
+    # read from available settings sources
+    if yaml_file is not None:
+        yaml_settings = yaml.safe_load(yaml_file)
+    elif default_pudl_yaml and default_pudl_yaml.exists():
+        with default_pudl_yaml.open() as f:
+            yaml_settings = yaml.safe_load(f)
+    else:
+        yaml_settings = {}
+
+    env_var_mapping = {"PUDL_INPUT": "pudl_in", "PUDL_OUTPUT": "pudl_out"}
```

Comment:
I also restructured this code to:

1. get all available settings
2. override the settings in a defined source order

That seemed easier to read for me.